File: 0367-valid-perfect-square/0367-valid-perfect-square.js
/**
 * @param {number} num
 * @return {boolean}
 */
var isPerfectSquare = function(num) {
    let [l, r] = [1, num];
    let pivot = Math.floor((l + r) / 2);
    while (l <= r) {
        if (pivot * pivot === num) {
            return true;
        } else if (pivot * pivot < num) {
            l = pivot + 1;
        } else {
            r = pivot - 1;
        }
        pivot = Math.floor((l + r) / 2);
    }
    return false;
};

File: 0367-valid-perfect-square/0367-valid-perfect-square.rb
# @param {Integer} num
# @return {Boolean}

def is_perfect_square(num)
    l, r = 1, num
    pivot = (l+r) / 2
    while l <= r
        return true if pivot*pivot == num
        if pivot*pivot < num
            l = pivot+1
        else
            r = pivot-1
        end
        pivot = (l+r) / 2
    end
    return false
end
